add HEFT gg_bb to the repo (with additional fixes for underflow FPEs)

valassi opened the pull request:
This is a WIP PR that logically follows the SUSY and SMEFT PRs. In particular it includes #632 which includes #824.

The main change is that I added heft_gg_bb to the repo (and am removing heft_gg_h).

Concerning heft_gg_bb, this now includes 4 diagrams (3 SM and also 1 HEFT) after fixing #828. This essentially means adding one diagram even if it is suppressed.

Now, the fact that the contribution is suppressed is probably why I start getting FPE underflows #831. I have now also fixed this (essentially by flushing to 0 all values of jamp_sv below E-18, so that their square is still above E-36 and fits in a float.

This is still WIP as I am running a few more tests.

valassi commented:
I have finally fixed also #831 FPEs for a variety of use cases. Now I am rerunning large scale tests, then this should also be ready to go.

Note, the LHE file mismatch #833 in heft_gg_bb will normally not be fixed in this MR.

valassi commented:
I had forgotten to backport the very last fix for #831, now it's done.

@oliviermattelaer I assign this to you as reviewer too, can you please have a look?

In the meantime I will run more tests, but this time hopefully the code should no longer change. Thanks!

valassi commented:
Merged here 'upstream/master' (with SUSY #824 and SMEFT #632)

oliviermattelaer commented on epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/madgraph/iolibs/template_files/gpu/cudacpp.mk:
```diff
@@ -833,7 +833,8 @@ endif
-$(testmain): LIBFLAGS += -L$(GTESTLIBDIR) -lgtest -lgtest_main
+$(testmain): LIBFLAGS += -L$(GTESTLIBDIR) -lgtest
+###$(testmain): LIBFLAGS += -lgtest_main # no longer necessary since we added main() to testxxx.cc
 
```
@oliviermattelaer  need to run on my mac to check if it still compiles

valassi replied:
Ok please check on your Mac, I suggest we go ahead and in case we fix later. But I do not see why it should fail: it is taking LESS from gtest, not more (main is now hardcoded in mg5amc). Resolving

oliviermattelaer reviewed (changes requested):
So we have some work to do but this is in quite good shape.

Thanks Andrea,

Olivier

valassi commented on epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/madgraph/iolibs/template_files/gpu/process_cc.inc:
```diff
@@ -40,5 +42,45 @@
+  //std::cout << "fpeEnable:     FE_OVERFLOW is" << ( ( fpes & FE_OVERFLOW ) ? " " : " NOT " ) << "enabled" << std::endl;
+  //std::cout << "fpeEnable:     FE_UNDERFLOW is" << ( ( fpes & FE_UNDERFLOW ) ? " " : " NOT " ) << "enabled" << std::endl;
+  const char* enableFPEc = getenv( "CUDACPP_RUNTIME_ENABLEFPE" );
+  const bool enableFPE = ( enableFPEc != 0 ) && ( std::string( enableFPEc ) != "" );
```
Hi @oliviermattelaer thanks. In the end I have disabled CUDACPP_RUNTIME_ENABLEFPE completely. I have simply enabled by default SIGFPE for the three cases invalid, divbyzero, overflow. So I think no need to add it to the runcards. Resolving.

valassi commented on epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/madgraph/iolibs/template_files/gpu/process_matrix.inc:
```diff
@@ -7,6 +7,32 @@
 
+#if not defined MGONGPUCPP_GPUIMPL and defined MGONGPU_FPTYPE2_FLOAT
+      // Avoid underflow FPE #831 for jamp (flush-to-zero jamp values whose square is below FLT_MIN)
+      constexpr fptype2 minjamp = 1.1 * constexpr_sqrt( FLT_MIN );
```
> if underflow occurs, it is not an issue (and even good) if we keep some printout to warn the user.

This is done. I have added printouts if underflow occurs in cudacpp.

> Could we change this in two ways:
> 
>     1. use a relative threshold compare to the maximum of the real/imaginary value of the jamp_sv
> 
>     2. put the threshold related to the numerical error of double precision (so around 1e-15)
> 
> 
> And we can made the code safer by
> 
>     1. one can rescale all jamp_sv by dividing by the max
> 
>     2. remultiply the final ztemp by the max square

Conversely, I have completely removed the code, for various reasons
- One, the new general strategy, after discussing with you, is do not try to prevent underflows, just let them happen and report them
- Two, after some tests, I realised that even FLT_MIN/10 is actually printed out to a reasonable value. I think the underflow just produces a 'denormal' with less precision than a float, but even usable, for some underflows. I prefer to just keep the code simpler at this stage.

Resolving

valassi commented:
Hi @oliviermattelaer thanks a lot for the discussion and review.

I made various changes acoording to what we discussed. Essentially the "new strategy for FPES #831" is the following:

- Enable SIGFPE by default for INVALID, DIVBYZERO, OVERFLOW via feeenableexcept. These are the FPEs that I extensively fixes in the ixx/oxx functions in the last few months anyway, so I hope there are none left: and if there are, we'd better fix them. I disabled (commented out) the env variable CUDACPP_RUNTIME_ENABLEFPE, now these three SIGFPES are enabled by default.

- About underflow (#831), which motivated many of the changes in this PR, in the end I decided that we should not treat it as a problem, and we just let it happen. So I removed this from the feenableexcept call below (it used to have fours FPEs, now it only has three). See also https://stackoverflow.com/questions/44308577/ieee-underflow-flag-ieee-denormal-in-fortran-77
and https://gcc.gnu.org/onlinedocs/gfortran/Debugging-Options.html. 

- By the way, about underflows I did some tests (committed in tools/fpe). Given FLT_MIN = 1.17549e-38, computing FLT_MIN / 10 does signal an underflow, but it prints out 1.17549e-39, not zero. I think (may be wrong) that these numbers are "denormals", with less precision than a float, but still more or less usable. So one more reason not to just flush-to-zero as I was doing.

- For all four FPEs above, I added a printout at the end if they have been reported or not. Normally this will only report underflows (because the other three will crash by design with SIGFPE, instead). Indeed in heft_gg_bb tests with float/mixed precision I did see underflows.

I am now doing a few more tests and cleanup, but I think this should be ok. I will ask you to review again when I am done. Thanks

valassi commented:
Hi @oliviermattelaer I confirm my changes are in. I re-request the review then (it should be easier now, a lot of code has disappeared!) Thanks

valassi commented:
Hi @oliviermattelaer I confirm I have finished. This is good to merge for me after your review. Thanks

valassi commented:
> For me this is good to go, maybe one detail would be to remove the special case for 'mdl_G__exp__2' and similar but that's ok if you want to keep it.

Hi Olivier, thanks a lot.

I had a look at the 'mdl_G__exp__2' and 'mdl_G__exp__3' issue. It seems that your patch fixed 'mdl_G__exp__3' but NOT 'mdl_G__exp__2'. I will add a comment in https://github.com/mg5amcnlo/mg5amcnlo/issues/89.

So anyway here I have removed my cudacpp patch for G^3 but kept it for G^2. I regenerated all code and am realunching the CI. I will merge when the CI is ok.

I will reopen https://github.com/mg5amcnlo/mg5amcnlo/issues/89 so that you can have a look. If/when you create a new patch I will test it in a separate madgraph4gpu PR.

Thanks again Andrea

valassi commented:
There are some issues on Mac see #838, I am trying to debug them and work around them

valassi commented:
My workaround for #838 was ok on ee_mumu. I am applying it on CODEGEN and all processes and relaunching the CI

valassi commented:
Ok the MacOS tests succeed after disabling testmisc (workaround for #838 - real fixes will be needed in #839 and #840).

All CI tests succeed. I am merging this. Thanks again @oliviermattelaer for the review and the approval.

(I still need to add a comment on https://github.com/mg5amcnlo/mg5amcnlo/issues/89, I will do that now. I got distracted bythe MacOS issues).
